Fix UB, panic and compiler error in `bevy_tasks`

NthTensor opened the pull request:
# Objective

Fixes several issues with `bevy_tasks` uncovered by @IceSentry and @Ratysz:
+ For web builds, `Task::is_finished` is missing, causing compile failures for cross-platform apps.
+ Scopes running on the single-threaded executor have possible UB due to mistakenly using a `'env` bound instead of a `'scope` lifetime bound on the scope ref.
+ Scopes running on the single-threaded executor fail to poll tasks spawned on them to completion when all tasks go to sleep at the same time (causing panics and potential UB).

## Solution

- Replace `futures_channel::oneshot` with `async_channel::bounded(1)`, and use `Receiver::is_empty` to implement `Task::finished`. The `async_channel` channel crate was chosen because it allows `Task` to be `Send + Sync + Unpin` and because it has a special-case oneshot implementation for bounded size-one channels (inherited from `concurrent-queue`). 
- Switch around the incorrect lifetime bounds.
- Use referencing counting to spin while ticking the executor until all tasks spawned on a scope are complete.

Since these are all fairly small changes I rolled them together into a single PR. I can split any of the three out if they prove controversial. 

## Testing

- Changes are currently untested. I need to check that this fixes the bugs we've been seeing at work, and I'd like to run this through miri.

It's concerning to me that miri failed to catch the bad lifetime bound on the scope impl.

MalekiRe reviewed (approved):
Lgtm

NthTensor commented:
Looks like `EdgeExecutor` is going to make this annoying.

NthTensor commented on crates/bevy_tasks/src/single_threaded_task_pool.rs:
```diff
@@ -167,8 +167,10 @@ impl TaskPool {
-        while executor.try_tick() {}
+        // Loop until all tasks are complete
+        while !executor.is_empty() {
+            block_on(executor.tick());
```
Hmm, I guess if you have a scoped-task that blocks on a non-scoped task, it could deadlock. Scoped-task depending on scoped-task is fine though.

I'll write a test-case for this. Might need to look closer at what the multi-threaded version does.

hymm replied:
> won't this deadlock if one task depends on another? since it wait for completion on only one task and won't poll the other one?

ah right. I was think awaiting would await for the task to complete, but it's just awaits for any task.

> Hmm, I guess if you have a scoped-task that blocks on a non-scoped task, it could deadlock. Scoped-task depending on scoped-task is fine though.

wouldn't this logic deadlock if there's a long running non scoped task running? i.e. a task that always returns pending.

NthTensor replied:
> wouldn't this logic deadlock if there's a long running non scoped task running? i.e. a task that always returns pending. So the executor is never empty.

That wouldn't be running on the scope's executor though. Right?

NthTensor replied:
I think part of the problem here is the [scoped tasks trilema](https://without.boats/blog/the-scoped-task-trilemma/). What `Scope` provides for the multi-threaded environment is `Parallelizability + Borrowing`. For the single-threaded version, it needs to provide `Concurrency + Borrowing` instead... I don't know if that's actually possible to do correctly using the same api.

james7132 commented:
> For web builds, Task::is_finished is missing, causing compile failures for cross-platform apps.

I'm surprised this was not caught by CI. Are we somehow skipping web builds?

NthTensor commented:
I circulated this at work have have gotten some internal feedback. Working on some fixes, and marking this as `waiting-on-author` for the moment.

NthTensor commented:
Pushed a change that should fix some of the missing trait bounds on wasm tasks (`Send + Sync + Unpin`). I've checked the docs and wasm tasks now implement exactly the same traits as the ones from `async_tasks`.

aevyrie commented on crates/bevy_tasks/Cargo.toml:
```diff
@@ -72,7 +72,7 @@ crossbeam-queue = { version = "0.3", default-features = false, features = [
 wasm-bindgen-futures = { version = "0.4", optional = true }
 pin-project = { version = "1", optional = true }
-futures-channel = { version = "0.3", optional = true }
+async-channel = { version = "2.5.0", optional = true }
```
2.3.0 is used elsewhere in tree

aevyrie replied:
Actually, this is *already* an option adept on line 64. I don't think this line is needed at all.

NthTensor replied:
Woops! Nice that we can converge on this impl for more things.

aevyrie replied:
Quick, add a "removes dependency" label to the PR

NthTensor commented:
I wrote a new implementation of `scope` for the single threaded case, and then almost immediately realized why the old version was implemented the way it was: any correct implementation might cause the main thread to block, which would either freeze or crash the page.

I've created a tracking issue here https://github.com/bevyengine/bevy/issues/20404.

NthTensor commented:
Pushed a prototype that merges the two different task implementations (in an effort to force us to have a consistent API) and adds reference counting to the single-threaded scope.

NthTensor commented:
This is now ready for review again, but I think we should probably try to get https://github.com/bevyengine/bevy/pull/20331 merged first.

james7132 commented on crates/bevy_tasks/src/task.rs:
```diff
@@ -36,25 +77,134 @@ impl<T> Task<T> {
-        Pin::new(&mut self.0).poll(cx)
+    cfg::web! {
+        if {
+            fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
```
nit: The function signature remains the same between both implementations, it can sit outside the cfg::web block.

NthTensor replied:
Signature isn't the same, self is `mut` in one of them. This avoids a clippy lint.

james7132 commented on crates/bevy_tasks/src/task.rs:
```diff
@@ -36,25 +77,134 @@ impl<T> Task<T> {
     }
 }
 
+// Tasks are futures that can be polled
```
All futures can be polled. Not sure what this comment is for.

NthTensor replied:
Yeah no idea what I was thinking there. I'll remove that.

james7132 commented on crates/bevy_tasks/src/task.rs:
```diff
@@ -36,25 +77,134 @@ impl<T> Task<T> {
     /// canceling because it also waits for the task to stop running.
     ///
-    /// See `async_executor::Task::cancel`
+    /// Canceling tasks is unsupported on the web, and this is the same as awaiting the task.
```
```suggestion
    /// # Platform-Specific Behavior
    /// Canceling tasks is unsupported on the web, and is the same as awaiting the task.
```

james7132 commented on crates/bevy_tasks/src/task.rs:
```diff
@@ -12,20 +15,58 @@ use core::{
 
-    /// Detaches the task to let it keep running in the background. See
-    /// `async_executor::Task::detach`
+// Main API
```
```suggestion
```

james7132 commented on crates/bevy_tasks/src/task.rs:
```diff
@@ -12,20 +15,58 @@ use core::{
 /// more gracefully and wait until it stops running, use the [`Task::cancel()`] method.
 ///
 /// Tasks that panic get immediately canceled. Awaiting a canceled task also causes a panic.
-#[derive(Debug)]
```
This is an API surface regression. Either keep the derive, or manually implement `Debug`.

NthTensor replied:
There's a manual debug impl lower down in the file. Derives don't work on type macros.

cBournhonesque commented on crates/bevy_tasks/src/single_threaded_task_pool.rs:
```diff
@@ -295,21 +305,32 @@ impl<'scope, 'env, T: Send + 'env> Scope<'scope, 'env, T> {
-                    *lock = Some(temp_result);
-                }
-            }
+            let result = f.await;
```
I'm not super familiar with async; but if `f` panics, then `pending_tasks` wouldn't get decremented and the scope would never return, no?

Or is that not a problem because this is single-threaded so we don't catch_unwind panics?

NthTensor replied:
I've changed it so that the scope uses it's own executor. This means panics will always propagate, as far as I can tell.

cBournhonesque commented on crates/bevy_tasks/src/task.rs:
```diff
@@ -36,25 +76,134 @@ impl<T> Task<T> {
+// All variants of Task<T> are expected to implement Unpin
+impl<T> Unpin for Task<T> {}
+
+// Derive dosn't work for macro types, so we have to implement this manually.
```
```suggestion
// Derive doesn't work for macro types, so we have to implement this manually.
```

cBournhonesque commented on crates/bevy_tasks/src/task.rs:
```diff
@@ -36,25 +76,134 @@ impl<T> Task<T> {
+                            unreachable!("catching a panic is only possible with std")
+                        }
+                    }},
+                    Poll::Ready(Err(_)) => panic!("Polled a task after it was cancelled"),
```
Isn't cancellation impossible in web?
Or I guess this could happen if we are in no-std and the web task panics, which means that the channel gets closed? If that's the case, maybe the panic message could be updated for clarity

NthTensor replied:
This code was just moved from `wasm_task.rs`, none of this logic is now. The only time this will error is if we get a `async_channel::RecvError` which means the sender-side is closed and the channel is empty, which means the task has already run and been awaited.

So it's correct to panic here. The message could be better though. It's trying to juggle the `async_tasks` terminology as best it can.

NthTensor replied:
Do you have a suggestion for a better error message?

NthTensor replied:
Done.

NthTensor commented on crates/bevy_tasks/src/single_threaded_task_pool.rs:
```diff
@@ -141,42 +149,42 @@ impl TaskPool {
+            results
+                .take()
+                .into_iter()
+                .map(|result| result.unwrap())
```
Yep, I see the issue here. If this function panics, any remaining scoped tasks spawned on the executor might have invalid refs, and so ticking the executor becomes UB. 

It would be easy to abort here. But since we don't use wasm-bindgen for scopes, we could actually cancel these tasks.

NthTensor replied:
I've moved back to using a fresh executor for each scope. The executor is dropped on a panic, so this is no longer an issue.

NthTensor commented on crates/bevy_tasks/src/single_threaded_task_pool.rs:
```diff
@@ -2,25 +2,33 @@ use alloc::{string::String, vec::Vec};
+        
+        struct Wrapper<T>(T);
+
+        #[expect(unsafe_code, reason = "hacking around nonsend bounds")]
```
I'll add one. The gist is, without `std` there should be no way to create threads. This makes the `Send` and `Sync` traits meaningless.

james7132 replied:
We may need to revisit that idea that Send/Sync is meaningless for no_std. I recently ran across [this comment](https://github.com/rust-lang/rust/issues/118959#issuecomment-2477823197) regarding this exact topic. While I definitely do not expect Bevy to ever support running in bare metal embedded spaces, I don't think we should be using such a justification for potentially unsound code, since those assumptions may break on some new no_std platform Bevy does want to support. Preliminary support for the Nintendo Switch, for example, was no_std, and it definitely could use multiple threads of execution.

NthTensor replied:
After thinking about this, I walked back some of my changes. We can continue to use `Executor` instead of `LocalExecutor` in the single-threaded static, and create a new `LocalExecutor` for every scope. This seems to satisfy the safety concerns here, and reduces the amount of unsafe code needed.

NthTensor commented on crates/bevy_tasks/src/lib.rs:
```diff
@@ -71,32 +71,35 @@ use alloc::boxed::Box;
+pub use futures_lite::future::poll_once;
 
+cfg::web! {
+    if {} else {
```
In this case, no.

NthTensor commented:
I've addressed the feedback from the latest reviews. Most notably: 
+ I've chosen to be less aggressive with refactoring the existing code.
+ I've chosen to use a fresh executor for every scope, which mostly eliminates the UB concern.
+ I've switched back from wrapping `LocalExecutor` to be send to just `Executor`.
+ I've added a test for the scope issue. It crashes on main but passes here.

james7132 commented on crates/bevy_tasks/src/single_threaded_task_pool.rs:
```diff
@@ -161,21 +165,15 @@ impl TaskPool {
-        // Loop until all tasks are done
-        while executor.try_tick() {}
+        // Wait until the socpe is complete
+        while *pending_tasks.borrow() != 0 {
+            block_on(executor.tick());
+        }
```
```suggestion
        block_on(executor.run(async  [
            while *pending_tasks.borrow() != 0 {
                future::yield_now().await;
            }
        }));
```
The current tick loop locks and unlocks a `Mutex` on each tick. The above code should avoid it. Is this feasible as an alternative?

NthTensor replied:
This seems to work.

james7132 reviewed (approved):
LGTM, though there's two things worth trying to see if we could improve the performance/readability here.

james7132 commented on crates/bevy_tasks/src/single_threaded_task_pool.rs:
```diff
@@ -161,21 +165,15 @@ impl TaskPool {
 
```
Nevermind, I forgot that that 'scope was HRTB'ed.

NthTensor commented:
Alright, last round of feedback addressed. As follow up to this, I want to investigate integrating `async_task` directly with `web_bindgen_futures`. But that's a longer term initiative.

I think this should be good for merge now, but I'm going to go ahead and re-request reviews.
